pisces/models.py

```diff
@@ -128,106 +41,95 @@ def predict_probabilities(self, sample_X: np.ndarray | pl.DataFrame) -> np.ndarr
+        self.data_processor = data_processor
 
-        y_demasked = np.where(y_trimmed < 0, 0, y_trimmed)
+    def get_needed_X_y(self, id: str) -> Tuple[np.ndarray, np.ndarray] | None:
+        return self.data_processor.get_1D_X_y(id)
 
-        return y_demasked, sample_weights
+    def train(self, 
+              examples_X: List[np.ndarray]=[], 
+              examples_y: List[np.ndarray]=[], 
+              pairs_Xy: List[Tuple[np.ndarray, np.ndarray]]=[],
+              epochs: int = 10,
+              ):
+        """
+        Assumes data is already preprocessed using `get_needed_X_y` 
+        and ready to be passed to the model.
 
-    def train(self, examples_X: List[pl.DataFrame] = [], examples_y: List[pl.DataFrame] = [], 
-              pairs_Xy: List[Tuple[pl.DataFrame, pl.DataFrame]] = [], 
-              epochs: int = 10, batch_size: int = 32):
-        if examples_X or examples_y:
-            assert len(examples_X) == len(examples_y)
+        Returns the loss history of the model.
+        """
+        if (examples_X and not examples_y) or (examples_y and not examples_X):
+            raise ValueError("If providing examples, must provide both X and y")
+        else:
+            if examples_X and examples_y:
+                assert len(examples_X) == len(examples_y)
         if pairs_Xy:
             assert not examples_X
-        
-        training = []
-        training_iterator = iter(pairs_Xy) if pairs_Xy else zip(examples_X, examples_y)
-        for X, y in training_iterator:
-            try:
-                X_folded = self._fold(X)
-                y_prepped, sample_weights = self._prepare_labels(y)
-                if (X_folded.shape[0] != y_prepped.shape[0]) \
-                    or (X_folded.shape[0] == 0) \
-                    or (y_prepped.shape[0] == 0):
-                    continue
-                training.append((X_folded, y_prepped, sample_weights))
-            except Exception as e:
-                print(f"Error folding or trimming data: {e}")
-                continue
-        
-        Xs = [X for X, _, _ in training]
-        ys = [y for _, y, _ in training]
-        weights = [w for _, _, w in training]
-        Xs = np.concatenate(Xs, axis=0)
-        ys = np.concatenate(ys, axis=0)
-        weights = np.concatenate(weights, axis=0)
+            examples_X = [pair[0] for pair in pairs_Xy]
+            examples_y = [pair[1] for pair in pairs_Xy]
+
+
+        Xs = np.concatenate(examples_X, axis=0)
+        ys = np.concatenate(examples_y, axis=0)
 
         selector = ys >= 0
         Xs = Xs[selector]
         ys = ys[selector]
-        weights = weights[selector]
 
-        self.pipeline.fit(Xs, ys)
+        # Get loss
+        old_stdout = sys.stdout
+        sys.stdout = mystdout = StringIO()
+        self.pipeline.fit(Xs, ys) # Fit the model
+        sys.stdout = old_stdout
+        loss_history = mystdout.getvalue()
+        loss_list = []
+        for line in loss_history.split('\n'):
```

Comment:
FWIW, parsing the logging messages of another package (or any logging messages) as part of your code is generally considered very fragile, because it requires the strings to be basically identical across subversions. 

But if the scikit-learn maintainers fix a typo in a string, or add a newline, that can wreck any downstream string parsing. From their perspective, though, it is kind of a trivial detail that may only bump the patch version because (hence should be compatible with any downstream code). 

That being said, I do like the idea of capturing the losses here in the way we can do elsewhere. I'm not seeing any built-in capabilities of sklearn, so maybe output parsing is as good as it gets for now. But in that case, we should wrap this parsing in a try-except block in case it fails.

Reply:
Good call! I was surprised that there isn't a way to obtain the loss history as an output. I think this works for now and definitely the try/except block makes it more robust!